src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs

```diff
@@ -703,12 +698,55 @@ private static FormulaValue Substitute(IRContext irContext, FormulaValue[] args)
                 instanceNum = (int)nv.Value;
             }
 
+            // Compute max possible memory this operation may need.
+            var sourceLen = source.Value.Length;
+            var matchLen = match.Value.Length;
+            var replacementLen = replacement.Value.Length;
+                        
+            int maxLenChars;
+            if (instanceNum < 0)
+            {
+                // Replace all instances. 
+                // Maximum possible length of Substitute, convert all the Match to Replacement. 
+                // Unicode, so 2B per character.
+                if (matchLen == 0)
+                {
+                    maxLenChars = sourceLen;
+                }
+                else
+                {
+                    maxLenChars = sourceLen / matchLen * replacementLen;
+                }
+            } 
+            else
+            {
+                // Only replace 1 instance 
+                maxLenChars = sourceLen + replacementLen;
+            }
+
+            var maxLenBytes = maxLenChars * 2;
+            runner.Governor.PollMemory(maxLenBytes);
+
+            var result = SubstituteWorker(runner, source, match, replacement, instanceNum);
+
+            return result; 
+        }
+
+        private static StringValue SubstituteWorker(EvalVisitor eval, StringValue source, StringValue match, StringValue replacement, int instanceNum)
+        {
+            if (string.IsNullOrEmpty(match.Value))
+            {
+                return source;
+            }
+
```

Comment:
Why is it moved inside the worker? No new memory would be allocated in this case.

Reply:
Every call to SubString() is an alloc!
This whole method needs to be rewritten  - #971.